accord-core/src/main/java/accord/topology/TopologyManager.java

```diff
@@ -0,0 +1,325 @@
+            this.syncTracker = new QuorumTracker(new Topologies.Singleton(global, false));
+            this.prevSynced = prevSynced;
+            updateState(prevSynced);
+        }
+
+        public boolean recordSyncComplete(Node.Id node, boolean prevSynced)
+        {
+            syncTracker.recordSuccess(node);
+            return updateState(prevSynced);
+        }
+
+        long epoch()
+        {
+            return global.epoch;
+        }
+
+        boolean syncComplete()
+        {
+            return prevSynced && syncComplete;
+        }
+
+        /**
+         * determine if sync has completed for all shards intersecting with the given keys
+         */
+        boolean syncCompleteFor(Keys keys)
+        {
+            if (!prevSynced)
+                return false;
+            if (syncComplete)
+                return true;
+            Boolean result = global.accumulateForKeys(keys, (i, shard, acc) -> {
+                if (acc == Boolean.FALSE)
+                    return acc;
+                return Boolean.valueOf(syncTracker.unsafeGet(i).hasReachedQuorum());
+            }, Boolean.TRUE);
+            return result == Boolean.TRUE;
+        }
+
+        boolean shardIsUnsynced(int idx, Shard shard)
+        {
+            return !syncTracker.unsafeGet(idx).hasReachedQuorum();
+        }
+    }
+
+    private class Epochs
+    {
+        private final long maxEpoch;
+        private final long minEpoch;
+        private final EpochState[] epochs;
+        private final List<Set<Node.Id>> pendingSyncComplete;
+
+        private Epochs(EpochState[] epochs, List<Set<Node.Id>> pendingSyncComplete)
+        {
+            this.maxEpoch = epochs.length > 0 ? epochs[0].epoch() : 0;
+            this.pendingSyncComplete = pendingSyncComplete;
+            for (int i=1; i<epochs.length; i++)
+                Preconditions.checkArgument(epochs[i].epoch() == epochs[i-1].epoch() - 1);
+            this.minEpoch = epochs.length > 0 ? epochs[epochs.length - 1].epoch() : 0;
+            this.epochs = epochs;
+        }
+
+        private Epochs(EpochState[] epochs)
+        {
+            this(epochs, new ArrayList<>());
+        }
+
+        public long nextEpoch()
+        {
+            return current().epoch + 1;
+        }
+
+        public Topology current()
+        {
+            return epochs.length > 0 ? epochs[0].global : Topology.EMPTY;
+        }
+
+        public Epochs add(Topology topology)
+        {
+            Preconditions.checkArgument(topology.epoch == nextEpoch());
+            EpochState[] nextEpochs = new EpochState[epochs.length + 1];
+            List<Set<Node.Id>> pendingSync = pendingSyncComplete;
+            if (!pendingSync.isEmpty())
+            {
+                boolean prevSynced = epochs.length <= 1 || epochs[1].syncComplete();
+                EpochState currentEpoch = epochs[0];
+                pendingSync.remove(0).forEach(id -> currentEpoch.recordSyncComplete(id, prevSynced));
```

Comment:
Why do we remove this element?

Reply:
`pendingSync` holds sync notifications we've received for future epochs (or the current epoch), and is indexed from the current epoch. Since we're adding an epoch here, the contents of pendingSync at index 0 are no longer for the current epoch, so they're removed from the list and applied to the epoch state.

Reply:
Thanks, I'm just confused why that translates into removing index 0 rather than pushing a new collection at index 0 and shifting the remainder down? If we're adding an epoch aren't we invalidating the whole sequence. I will be reading this all again more carefully today, hopefully with my brain working better, so hopefully I'll grok it properly.

Reply:
pendingSync is indexed with the current epoch at 0, plus epochs we haven't seen yet indexed after that. So if the current epoch is 5, the epochs represented in pending sync would be [5, 6, 7]. If we receive a new epoch topology (6), the pending sync entry representing epoch 5 must be removed for this indexing assumption to be valid